compiler/src/iree/compiler/Codegen/Common/CombineLayoutTransformation.cpp

```diff
@@ -198,12 +210,159 @@ foldExtractSliceIntoMapScatter(RewriterBase &rewriter,
+      rewriter.getArrayAttr(distConfig.mapping),
+      /*bodyBuilder=*/[&](OpBuilder &b, Location nestedLoc, ValueRange ivs) {
+        SmallVector<OpFoldResult> nestedLbs(ivs);
+        SmallVector<OpFoldResult> nestedUbs;
+        for (auto [start, step, ub] : llvm::zip_equal(nestedLbs, steps, ubs)) {
+          auto tileEnd = IREE::LinalgExt::addOfrs(b, nestedLoc, start, step);
+          auto minMap = AffineMap::get(
+              2, 0, {b.getAffineDimExpr(0), b.getAffineDimExpr(1)},
+              rewriter.getContext());
+          auto min = affine::makeComposedFoldedAffineMin(b, nestedLoc, minMap,
+                                                         {tileEnd, ub});
+          nestedUbs.push_back(min);
+        }
+        // Continue distribution if there are more distribution levels.
+        if (distributionLevel + 1 < distConfigs.size()) {
+          buildNestedDistributionLoops(
+              rewriter, nestedLoc, distributionLevel + 1, nestedLbs, nestedUbs,
+              distConfigs, innerLoopBuilder);
+          b.create<scf::InParallelOp>(nestedLoc);
+          return;
+        }
+        // Otherwise, tile to one, and generate the inner loop body.
+        SmallVector<Value> nestedLbVals =
+            getValueOrCreateConstantIndexOp(b, nestedLoc, nestedLbs);
+        SmallVector<Value> nestedUbVals =
+            getValueOrCreateConstantIndexOp(b, nestedLoc, nestedUbs);
+        Value one = rewriter.create<arith::ConstantIndexOp>(nestedLoc, 1);
+        SmallVector<Value> unitSteps(nestedLbs.size(), one);
+        scf::buildLoopNest(rewriter, nestedLoc, nestedLbVals, nestedUbVals,
+                           unitSteps, innerLoopBuilder);
+        b.create<scf::InParallelOp>(nestedLoc);
+      });
+}
+
+/// Fold a tensor.pad op into a iree_linalg_ext.map_scatter op, and separate
+/// the writing of padding values into a separate operation on the buffer that
+/// the map_scatter op is ultimately written into. The result buffer is taken
+/// from the direct consumer of the `mapScatterOp`, which is expected to be an
+/// `iree_codegen.store_to_memref` op. Return failure if the result buffer is
+/// not found.
+static FailureOr<MapScatterOp>
+foldPadIntoMapScatter(RewriterBase &rewriter, tensor::PadOp padOp,
+                      MapScatterOp mapScatterOp,
+                      DistributionConfigFn distributionConfigFn) {
+  // Find the output buffer that the mapScatterOp is stored into.
+  if (!mapScatterOp->hasOneUse()) {
+    return rewriter.notifyMatchFailure(
+        mapScatterOp, "map_scatter does not have a single user");
+  }
+  auto storeOp = dyn_cast<IREE::Codegen::StoreToMemrefOp>(
+      *mapScatterOp->getUsers().begin());
+  if (!storeOp) {
+    return rewriter.notifyMatchFailure(
+        mapScatterOp,
+        "map_scatter user is not an iree_codegen.store_to_memref op");
+  }
+  Value outputBuffer = storeOp.getTarget();
+
+  // Write the padding values directly into the outputBuffer.
```

Comment:
The comment seems to fit the `innerLoopBuilder` better. Maybe move the comment right before `innerLoopBuilder` and the `outputBuffer` is closer to where it is used. I think you declare the `outputBuffer` here just because of the comment?

```cpp
  // Write the padding values directly into the outputBuffer.
  Value outputBuffer = storeOp.getTarget();
  auto innerLoopBuilder = [&](OpBuilder &b, Location loopLoc, ValueRange ivs) {
```